qadence/backends/pyqtorch/convert_ops.py

```diff
@@ -350,13 +362,19 @@ def _hamiltonian(self: PyQHamiltonianEvolution, values: dict[str, Tensor]) -> Te
         self._dtype: torch_dtype = self.hmat.dtype if hasattr(self, "hmat") else cdouble
 
     def _unitary(self, hamiltonian: Tensor, time_evolution: Tensor) -> Tensor:
-        self.batch_size = max(hamiltonian.size()[2], len(time_evolution))
-        diag_check = tensor(
-            [is_diag(hamiltonian[..., i]) for i in range(hamiltonian.size()[2])], device=self.device
-        )
+
+        if len(hamiltonian.size()) > 1:
+            self.batch_size = max(hamiltonian.size()[2], len(time_evolution))
+            diag_check = tensor(
+                [is_diag(hamiltonian[..., i]) for i in range(self.batch_size)], device=self.device
+            )
+            hamiltonian = diagonal(hamiltonian) if bool(diag_check) else hamiltonian
+        else:
+            self.batch_size = None
+            diag_check = tensor([True])
 
         def _evolve_diag_operator(hamiltonian: Tensor, time_evolution: Tensor) -> Tensor:
-            evol_operator = diagonal(hamiltonian) * (-1j * time_evolution).view((-1, 1))
+            evol_operator = hamiltonian * (-1j * time_evolution).view((-1, 1))
             evol_operator = diag_embed(exp(evol_operator))
             return transpose(evol_operator, 0, -1)
```

Comment:
The logic here needs to be reworked a bit, and probably requires a bit of a refactor. There is no longer a need to use the `is_diag` imported from PyQTorch because you already checked for diagonality with the `block.generator._is_diag_pauli`. It seems to me like the `_evolve_diag_operator` needs to be decoupled from the `_unitary` method somehow. The point is that we need to avoid doing  `diag_embed(exp(evol_operator))` because that will instatiate a dense full matrix from the diagonal, which we don't want. We want to use the diagonal vector directly when applying it to the state through a vector - vector product.

Reply:
@jpmoutinho, yes I agree. 
In case of diagonal hamiltonian, we have 1-d vector representation, where evolution operator can be simply calculated as exponent function and applied to specific state through multiplication, in other way the state will be evolved using the function` pyq.apply_operator`.
This logic can be implemented in the `self.forward` function.